### PR TITLE
`CallBuilder` delegated calls API

### PR DESCRIPTION
**Pull request by VargSupercolony**
This PR adds the possibility to make delegated calls into ink!. To achieve this, we:
1. Add `seal_delegate_call` unstable interface import
2. Refactor `CallBuilder` to two types of calls: `Call` (meaning cross-contract call) and `DelegateCall` (meaning delegated calls). This is a breaking change, as now the users must specify `call_type` on the builder manually, as opposed to it being always `Call` :)
3. Add `invoke_contract_delegate` & `eval_contract_delegate` functions to backend trait & impls: we call those functions when making a delegated call
4. Rework codegen for cross-contract calls (add `set_call_type`)
5. Update docs in the ink! itself
6. Update `proxy` example, rework it to use `build_call` with `DelegateCall` call type, add upgradeable flipper to showcase simple upgradeable contract example.

Massive thanks to @yarikbratashchuk for providing initial PoCs & ideas for the implementation.

### CODE REVIEW

File: crates/env/src/engine/off_chain/impls.rs [hunk 1]
```diff
@@ -420,9 +426,28 @@ impl TypedEnvBackend for EnvInstance {
         unimplemented!("off-chain environment does not support contract invocation")
     }
 
+    #[allow(clippy::type_complexity)]
```
Comment (xgreenx) [inline crates/env/src/engine/off_chain/impls.rs:429]: Maybe we can create an alias to avoid `type_complexity` everywhere=)

File: examples/proxy/lib.rs [hunk 2]
```diff
@@ -9,53 +9,93 @@
...
+    use ink_env::call::DelegateCall;
+    use ink_primitives::{
+        Key,
+        KeyPtr,
+    };
+    use ink_storage::traits::SpreadLayout;
+
+    #[derive(Debug)]
+    #[cfg_attr(feature = "std", derive(ink_storage::traits::StorageLayout))]
+    struct ProxyFields {
```
Comment (xgreenx) [inline examples/proxy/lib.rs:28]: Could you add a comment with the description "Why did you move all fields to separate structure"?

File: crates/env/src/engine/off_chain/impls.rs (hunk 1 above)
Comment (VargSupercolony) [reply crates/env/src/engine/off_chain/impls.rs:429]: Nope, the compiler gives a [warning](https://stackoverflow.com/questions/37871267/is-it-possible-to-create-a-type-alias-that-has-trait-bounds-on-a-generic-type-fo) that the type bounds are not checked.

**Comment by athei** [comment]
Can you please satisfy the CI?

**Comment by VargSupercolony** [comment]
@athei I will fix the errors I get now, but I remember some tests were failing in the `ink-waterfall` testing job. Is it fixable by doing fixes into my PR, or are there any external tests that are run not from ink?

**Comment by athei** [comment]
The ink waterfall doesn't even start because it is failing before it comes to it. `ink-waterfall` only runs if the basic tests are successful.

`ink-waterfall` will fail because the node is too old and you added new functionality to examples which are deployed by the waterfall.  I will upgrade the node it so it will be successful tomorrow (waterfall uses latest `master` of `substrate-contracts-node` but only updated nightly).

For the new examples you added we would need manually ad them to the `ink-waterfall` so that they are also deployed and testes there. But first we need to fix CI here.

**Comment by VargSupercolony** [comment]
@athei, please see that the CI completes before I update to the latest master.

File: crates/env/src/call/call_builder.rs [hunk 3]
```diff
@@ -210,118 +295,172 @@ where
...
 
-/// Builds up a cross contract call.
-pub struct CallBuilder<E, Callee, GasLimit, TransferredValue, Args, RetType>
-where
-    E: Environment,
-{
+/// The default call type for cross-contract calls. Performs a cross-contract call to `callee`
+/// with gas limit `gas_limit`, transferring `transferred_value` of currency.
+pub struct Call<E, Callee, GasLimit, TransferredValue> {
     env: PhantomData<fn() -> E>,
```
Comment (athei) [inline crates/env/src/call/call_builder.rs:308]: Like above should be called `_phantom`.

**Comment by cmichi** [comment]
@VargSupercolony Thanks for the PR. The `ink-waterfall` CI stage does not succeed because you modified the `flipper` example and the waterfall tests this with some defined interactions, which are no longer valid with your PR.

I would prefer if you leave the existing `proxy` example intact and instead create a new one `upgradeable-contract`. Could you do that please?

**Comment by cmichi** [comment]
@VargSupercolony We're shaping up for a stable ink! release ‒ hopefully this week. It would be awesome to have this PR in there! Do you think you can implement the comments this week?

**Comment by armyhaylenko** [comment]
hey @cmichi @athei, writing from my personal acc as I don't have much internet access. I will not be able to work with the PR in the coming days because of war, and I parted my ways with Supercolony, so I think someone will pick it up - probably @xgreenx :) but anything you need from me - please mention this acc, this is my personal one

**Comment by xgreenx** [comment]
> @VargSupercolony We're shaping up for a stable ink! release ‒ hopefully this week. It would be awesome to have this PR in there! Do you think you can implement the comments this week?

> hey @cmichi @athei, writing from my personal acc as I don't have much internet access. I will not be able to work with the PR in the coming days because of war, and I parted my ways with Supercolony, so I think someone will pick it up - probably @xgreenx :) but anything you need from me - please mention this acc, this is my personal one

Yes, I will work on it, I will try to update it today=)

File: examples/proxy/lib.rs [hunk 4]
```diff
@@ -9,53 +9,99 @@
...
+            SpreadLayout::push_spread(&self.delegate_to, &mut ptr);
+            SpreadLayout::push_spread(&self.admin, &mut ptr);
+        }
+
+        fn clear_spread(&self, _: &mut KeyPtr) {
+            let mut ptr = KeyPtr::from(Key::from(PROXY_FIELDS_STORAGE_KEY));
+            SpreadLayout::clear_spread(&self.delegate_to, &mut ptr);
+            SpreadLayout::clear_spread(&self.admin, &mut ptr);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (xgreenx) [reply examples/proxy/lib.rs:70]: Yep, but maybe it is okay for now because it will be fixed with https://github.com/paritytech/ink/issues/1134

**Comment by xgreenx** [comment]
@cmichi I returned the `proxy` example back, but ink-waterfall still fails=(

**Comment by cmichi** [comment]
> @cmichi I returned the `proxy` example back, but ink-waterfall still fails=(

I'm still looking into it, but AFAIU it has nothing to do with this PR, the tests actually all run through. The CI only complains when shutting down. I currently suspect some combination of protected variables and first time external contribution. 

Let's ignore the `ink-waterfall` tests here for now.

File: examples/upgradable-contract/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+[lib]
+name = "upgradable_contract"
+path = "lib.rs"
+crate-type = ["cdylib"]
+
+# Needed until https://github.com/paritytech/ink/issues/364 is resolved.
+[profile.release]
+overflow-checks = false
+
```
[the remark this answers is not in the document]
Comment (cmichi) [reply examples/upgradable-contract/Cargo.toml]: No longer needed since https://github.com/paritytech/ink/pull/1049.

File: examples/upgradable-contract/README.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+In order to test it out you need to do the following:
+
+1. Build a contract containing some logic, e.g. our upgradeable flipper example:
+   ```
+   cargo +nightly contract build --manifest-path=examples/upgradable-contract/upgradeable_flipper/Cargo.toml
+   ```
+   You will receive the respective `upgradeable_flipper.contract` bundle in the `examples/upgradable-contract/upgradeable_flipper/target/ink/` folder.
+
+   In order to perform migrations and have proxy working for contracts with different storage
+   layouts, we use the [`Upgradeable`](upgradeable_flipper/upgradeable.rs) type wrapper, which ensures
```
Comment (cmichi) [inline examples/upgradable-contract/README.md:20]: We have the term "upgradeable" and "upgradable" in this PR. @ascjones :gb: Wdyt?

Comment (xgreenx) [reply examples/upgradable-contract/README.md:20]: I will rename everything to `upgradeable`

**Comment by paritytech-cicd-pr** [comment]
## 🦑 📈 ink! Example Contracts ‒ Changes Report 📉 🦑
 ⚠️ **The ink! `master` is ahead of your branch, this might skew the comparison data below.**

These are the results when building the `examples/*` contracts from this branch with `cargo-contract 0.17.0-638ad57` and comparing them to ink! `master`: 

|                        | Δ Optimized Size | Δ Used Gas | Total Optimized Size | Total Used Gas |
|:------------------------|------------------:|------------:|----------------------:|----------------:|
| `accumulator`          |                  |            | 1.28 K               |                |
| `adder`                | +0.05 K          |            | 2.28 K               |                |
| `contract-terminate`   |                  |            | 1.23 K               | 214_418        |
| `contract-transfer`    |                  |            | 8.37 K               | 14_418         |
| `delegator`            | +0.01 K          | +145       | 6.25 K               | 46_649         |
| `dns`                  |                  |            | 9.62 K               | 43_254         |
| `erc1155`              |                  |            | 27.49 K              | 86_508         |
| `erc20`                |                  |            | 9.09 K               | 43_254         |
| `erc721`               |                  |            | 14.37 K              | 115_344        |
| `flipper`              |                  |            | 1.55 K               | 14_418         |
| `incrementer`          |                  |            | 1.49 K               | 14_418         |
| `multisig`             | +0.01 K          | +2         | 26.13 K              | 94_109         |
| `proxy`                | -0.06 K          | +30        | 3.17 K               | 29_593         |
| `rand-extension`       |                  |            | 4.38 K               | 14_418         |
| `subber`               | +0.05 K          |            | 2.29 K               |                |
| `trait-erc20`          |                  |            | 9.36 K               | 43_254         |
| `trait-flipper`        |                  |            | 1.33 K               | 14_418         |
| `trait-incrementer`    |                  |            | 1.42 K               | 28_836         |
| `upgradeable-contract` | +3.22 K          |            | 3.22 K               |                |



[Link to the run](https://gitlab.parity.io/parity/ink-waterfall/-/pipelines/181631) | Last update: Wed Mar  2 14:59:52 CET 2022

**Comment by cmichi** [comment]
I've found the issue which caused the `ink-waterfall` stage to fail!

(It was because this PR adds a new example, which is not yet tested in the waterfall)

File: examples/upgradeable-contract/upgradeable-flipper/upgradeable.rs [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+    fn allocate_packed(&mut self, at: &Key) {
+        <T as PackedAllocate>::allocate_packed(&mut self.inner, at)
+    }
+}
+
+impl<T: PackedAllocate> PackedAllocate for Upgradeable<T, NotInitialized> {
+    fn allocate_packed(&mut self, at: &Key) {
+        <T as PackedAllocate>::allocate_packed(&mut self.inner, at)
+    }
+}
```
Comment (cmichi) [inline examples/upgradeable-contract/upgradeable-flipper/upgradeable.rs:129]: :scream: Wow that's a lot of boilerplate code for the storage. @xgreenx Do you think this will reduce significantly with https://github.com/paritytech/ink/issues/1134 ?

Comment (xgreenx) [reply examples/upgradeable-contract/upgradeable-flipper/upgradeable.rs:129]: Of course, we only will use only one trait(maybe two), and that trait will be automatically implemented by `#[ink::storage_ext]`. All keys will be managed by the user during the definition of the struct, so the code should look tiny=)

Comment (athei) [reply examples/upgradeable-contract/upgradeable-flipper/upgradeable.rs:129]: Yeah all of this boilerplate is only needed because the current system doesn't allow to change the contract root key without manually implementing `SpreadLayout`.

File: examples/upgradable-contract/README.md (hunk 6 above)
Comment (HCastano) [reply examples/upgradable-contract/README.md]: We've had this debate over on the Substrate side, and American spelling should be used 🇺🇸

File: examples/upgradeable-contract/upgradeable-flipper/upgradeable.rs (hunk 7 above)
Comment (HCastano) [reply examples/upgradeable-contract/upgradeable-flipper/upgradeable.rs:129]: The boilerplate, oh my 🙈

**Review by HCastano — COMMENTED** [review]
Only took a quick peek today, will need to spend more time on it later in the week.

One thing that comes to mind is that we should really look for ways to simplify that example contract. If I were a newcomer to `ink!` and saw this as the base-case for upgradeable contracts I'd definitely be scared off

**Comment by xgreenx** [comment]
> One thing that comes to mind is that we should really look for ways to simplify that example contract. If I were a newcomer to ink! and saw this as the base-case for upgradeable contracts I'd definitely be scared off

With https://github.com/paritytech/ink/issues/1134 we can remove the `Upgradeable` struct. But we should introduce a new argument for `#[ink::storage]`. The problem is that in the proxy pattern the constructor of the callee contract is not called and storage is not initialized. It requires some workaround because nothing is stored under the storage key. For that, we created the `Upgradeable` wrapper. 

But with https://github.com/paritytech/ink/issues/1134 we need to provide a way how to init the contract's storae if is not initialized. For example, argument in macro `#[ink::storage(init = default())]` or `#[ink::storage(default_if_not_exist = true)]`

**Review by HCastano — COMMENTED** [review]
I'm like halfway through looking at this. Main thing I haven't look at yet are the examples

File: crates/env/src/call/call_builder.rs [hunk 8]
```diff
@@ -20,71 +20,87 @@ use crate::{
...
     }
 
+    /// Returns the execution input.
+    #[inline]
+    pub(crate) fn exec_input(&self) -> &ExecutionInput<Args> {
+        &self.exec_input
+    }
+}
+
+impl<E, Args, R> CallParams<E, Call<E, E::AccountId, u64, E::Balance>, Args, R>
```
[the remark this answers is not in the document]
Comment (xgreenx) [reply crates/env/src/call/call_builder.rs]: I tried to put the definition of the `Gas` into `Environment` , but it affects a lot of files, so better to od it in another PR=)

File: crates/env/src/api.rs [hunk 9]
```diff
@@ -242,6 +247,32 @@ where
     })
 }
 
+/// Invokes a contract message via delegate call.
+///
+/// # Note
+///
+/// - Prefer using this over [`eval_contract_delegate`] if possible. [`invoke_contract_delegate`]
```
[the remark this answers is not in the document]
Comment (ascjones) [reply crates/env/src/api.rs:255]: I think that would be a good idea, yes.

File: examples/upgradeable-contract/upgradeable-flipper/upgradeable.rs [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+};
+
+#[derive(Debug)]
+pub struct Initialized;
+#[derive(Debug)]
+pub struct NotInitialized;
+
+#[derive(Debug, Decode, Encode)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub struct Upgradeable<T: PackedLayout, InitializationStatus = Initialized> {
```
Comment (cmichi) [inline examples/upgradeable-contract/upgradeable-flipper/upgradeable.rs:24]: Could you add some doc comments to these `pub`'s and a comment on why the `PackedLayout` / `SpreadLayout` implementations all have to be implemented? I think something like the comment by Alex would already be sufficient:

>Yeah all of this boilerplate is only needed because the current system doesn't allow to change the contract root key without manually implementing SpreadLayout.

File: examples/upgradeable-contract/upgradeable-flipper/Cargo.toml [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+std = [
+    "ink_primitives/std",
+    "ink_metadata",
+    "ink_metadata/std",
+    "ink_env/std",
+    "ink_storage/std",
+    "ink_lang/std",
+    "scale/std",
+    "scale-info",
+    "scale-info/std",
```
Comment (cmichi) [inline examples/upgradeable-contract/upgradeable-flipper/Cargo.toml:33]:
```suggestion
    "ink_primitives/std",
    "ink_metadata/std",
    "ink_env/std",
    "ink_storage/std",
    "ink_lang/std",
    "scale/std",
    "scale-info/std",
```

File: examples/upgradeable-contract/upgradeable-flipper/Cargo.toml [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+[lib]
+name = "upgradeable_flipper"
+path = "lib.rs"
+crate-type = ["cdylib"]
+
+[features]
+default = ["std"]
+std = [
+    "ink_primitives/std",
+    "ink_metadata",
```
Comment (cmichi) [inline examples/upgradeable-contract/upgradeable-flipper/Cargo.toml:26]:
```suggestion
```

Comment (cmichi) [reply examples/upgradeable-contract/upgradeable-flipper/Cargo.toml:26]: Redundant because of the line below.

File: examples/upgradeable-contract/upgradeable-flipper/Cargo.toml [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+    "ink_metadata/std",
+    "ink_env/std",
+    "ink_storage/std",
+    "ink_lang/std",
+    "scale/std",
+    "scale-info",
```
Comment (cmichi) [inline examples/upgradeable-contract/upgradeable-flipper/Cargo.toml:32]:
```suggestion
```
Same here.

**Comment by HCastano** [comment]
@xgreenx can you avoid force-pushing PRs? It makes pulling the changes down easier, and also makes it harder to see what commits are new.

It's fine to merge master in order to keep up to date

File: crates/env/src/call/call_builder.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (HCastano) [reply crates/env/src/call/call_builder.rs]: Yeah, I thought about that as well, but also figured it would be messy to do here

**Comment by xgreenx** [comment]
> @xgreenx can you avoid force-pushing PRs? It makes pulling the changes down easier, and also makes it harder to see what commits are new.
>
> It's fine to merge master in order to keep up to date

I'm only forcing it to rebase to master.
During each merge with the master, I need to resolve conflicts=) That means without rebasing I need to resolve all previous conflicts again and a new one.

**Comment by xgreenx** [comment]
`ink-waterfall` fails because of a new test that is using builder=(
image

File: examples/upgradeable-contract/upgradeable-flipper/upgradeable.rs [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+    }
+}
+
+/// It is implementation of `SpreadLayout` that initialize the inner type if it is not initialized.
+impl<T: PackedLayout + SpreadAllocate> SpreadLayout for Upgradeable<T, NotInitialized> {
+    const FOOTPRINT: u64 = <T as SpreadLayout>::FOOTPRINT;
+    const REQUIRES_DEEP_CLEAN_UP: bool = <T as SpreadLayout>::REQUIRES_DEEP_CLEAN_UP;
+
+    fn pull_spread(ptr: &mut KeyPtr) -> Self {
+        if ink_env::get_contract_storage::<T>(ptr.advance_by(0))
```
Comment (HCastano) [inline examples/upgradeable-contract/upgradeable-flipper/upgradeable.rs:71]:
```suggestion
        if ink_env::get_contract_storage::<T>(ptr.key())
```

File: examples/upgradeable-contract/upgradeable-flipper/upgradeable.rs [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+    fn allocate_packed(&mut self, at: &Key) {
+        <T as PackedAllocate>::allocate_packed(&mut self.inner, at)
+    }
+}
+
+impl<T: PackedAllocate> PackedAllocate for Upgradeable<T, NotInitialized> {
+    fn allocate_packed(&mut self, at: &Key) {
+        <T as PackedAllocate>::allocate_packed(&mut self.inner, at)
+    }
+}
```
Comment (HCastano) [inline examples/upgradeable-contract/upgradeable-flipper/upgradeable.rs:142]: Can we not make these generic over `Upgradeable<T, S>` the implementations are the same for `Initialized` and `NotInitialized`?

File: examples/upgradeable-contract/upgradeable-flipper/upgradeable.rs [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+        &self.inner
+    }
+}
+
+impl<T: PackedLayout, State> AsMut<T> for Upgradeable<T, State> {
+    #[inline]
+    fn as_mut(&mut self) -> &mut T {
+        &mut self.inner
+    }
+}
```
Comment (HCastano) [inline examples/upgradeable-contract/upgradeable-flipper/upgradeable.rs:170]: I know these impls are nice to have, but it doesn't look like they're used anywhere. For the sake of reducing boilerplate in the example I'd remove them

**Review by HCastano — COMMENTED** [review]
In the right direction, but needs a bit of clean-up

File: crates/env/src/call/call_builder.rs [hunk 17]
```diff
@@ -20,71 +20,87 @@ use crate::{
...
+    pub(crate) fn callee(&self) -> &E::AccountId {
+        &self.call_type.callee
+    }
+
     /// Returns the chosen gas limit for the called contract execution.
     #[inline]
-    pub(crate) fn gas_limit(&self) -> u64 {
-        self.gas_limit
+    pub(crate) fn gas_limit(&self) -> &Gas {
+        &self.call_type.gas_limit
```
[the remark this answers is not in the document]
Comment (xgreenx) [reply crates/env/src/call/call_builder.rs:82]: To be consistent with code around)
image

File: crates/env/src/engine/on_chain/ext.rs [hunk 18]
```diff
@@ -338,6 +338,15 @@ mod sys {
             message_hash_ptr: Ptr32<[u8]>,
             output_ptr: Ptr32Mut<[u8]>,
         ) -> ReturnCode;
+
+        pub fn seal_delegate_call(
```
[the remark this answers is not in the document]
Comment (xgreenx) [reply crates/env/src/engine/on_chain/ext.rs:342]: It is in the "unstable" section. `seal_call` in "stable"

File: examples/upgradeable-contract/lib.rs [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+    ///
+    /// That allows to store the structure in separate cells and it will not conflict
+    /// with the default storage layout of the callee contract.
+    impl SpreadLayout for ProxyFields {
+        const FOOTPRINT: u64 =
+            <AccountId as SpreadLayout>::FOOTPRINT + <Hash as SpreadLayout>::FOOTPRINT;
+        const REQUIRES_DEEP_CLEAN_UP: bool = false;
+
+        fn pull_spread(_: &mut KeyPtr) -> Self {
+            let mut ptr = KeyPtr::from(Key::from(PROXY_FIELDS_STORAGE_KEY));
```
[the remark this answers is not in the document]
Comment (xgreenx) [reply examples/upgradeable-contract/lib.rs:55]: No, because rust doesn't allow `cosnt NAME: KeyPtr = KeyPtr::from(...)`.

File: examples/upgradeable-contract/upgradeable-flipper/upgradeable.rs (hunk 14 above)
Comment (xgreenx) [reply examples/upgradeable-contract/upgradeable-flipper/upgradeable.rs:71]: `key` returns the previous key, we need here an actual key, so we are forcing `advance_by(0)`.
image

File: examples/upgradeable-contract/upgradeable-flipper/upgradeable.rs (hunk 15 above)
Comment (xgreenx) [reply examples/upgradeable-contract/upgradeable-flipper/upgradeable.rs:142]: No, we can't, because `SpreadAllocate` requires `SpreadAllocate`. `SpreadAllocate` is implemented only for `Upgradeable<T, NotInitialized>` and `Upgradeable<T, Initialized>` so we can't create a generic implementation.

**Review by cmichi — APPROVED** [review]
Thanks!

The failing `ink-waterfall` CI is because it tests some additional ink! examples which don't make sense to be committed into the ink! repo, they don't use the new API yet. So the failing CI stage can be ignored here. It's this one: https://github.com/paritytech/ink-waterfall/blob/master/examples/contract-introspection/lib.rs.

File: examples/upgradeable-contract/lib.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (HCastano) [reply examples/upgradeable-contract/lib.rs:58]: So `Key` has a [constructor that's `const`](https://paritytech.github.io/ink/ink_primitives/struct.Key.html#method.new). We could add a similar one for `KeyPtr` to achieve this. But maybe let's do that in a different PR

File: crates/env/src/call/call_builder.rs [hunk 20]
```diff
@@ -199,131 +277,157 @@ where
...
-    /// The current parameters that have been built up so far.
-    callee: Callee,
-    call_flags: CallFlags,
-    gas_limit: GasLimit,
-    transferred_value: TransferredValue,
-    exec_input: Args,
-    return_type: RetType,
+/// The default call type for cross-contract calls. Performs a cross-contract call to `callee`
+/// with gas limit `gas_limit`, transferring `transferred_value` of currency.
+pub struct Call<E: Environment> {
```
Comment (HCastano) [inline crates/env/src/call/call_builder.rs:298]: Yeah I like this more actually 👍

File: crates/env/src/call/call_builder.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (HCastano) [reply crates/env/src/call/call_builder.rs:82]: I don't think it makes sense in this case. We just end up dereferencing `Gas` immediately anyways

File: crates/lang/src/env_access.rs [hunk 21]
```diff
@@ -35,12 +37,12 @@ use ink_eth_compatibility::ECDSAPublicKey;
 /// This allows ink! messages to make use of the environment efficiently
 /// and user friendly while also maintaining access invariants.
 #[derive(Copy, Clone)]
-pub struct EnvAccess<'a, T> {
-    /// Tricks the Rust compiler into thinking that we use `T`.
-    marker: PhantomData<fn() -> &'a T>,
+pub struct EnvAccess<'a, E> {
```
Comment (HCastano) [inline crates/lang/src/env_access.rs:40]: I get why you renamed `T` to `E`, but in the future can you at least split this out into its own commit? It adds a lot of unnecessary noise. Tbh it should've been done in a follow-up PR

File: examples/upgradeable-contract/README.md [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+   `examples/upgradeable-contract/upgradeable-flipper/target/ink/` folder.
+
+   In order to perform migrations and have proxy working for contracts with different 
+   [storage layouts](https://paritytech.github.io/ink-docs/datastructures/spread-storage-layout), 
+   we use the [`Upgradeable`](upgradeable-flipper/upgradeable.rs) type 
+   wrapper, which ensures that we write different fields of desired struct to different 
+   storage locations, while also tracking the initialization status (e.g., we uploaded 
+   the code on chain, but haven't called the constructor).
+
+3. Build the proxy contract:
```
Comment (HCastano) [inline examples/upgradeable-contract/README.md:27]: If you change all of these numbers to `1` Github is smart enough to render the the numbers in the correct order. Makes it easier when change are introduced (no need to update the entire list)

File: examples/upgradeable-contract/lib.rs [hunk 23]
```diff
@@ -0,0 +1,140 @@
...
+    /// The reason this is a separate structure is that we want to keep
+    /// the data for this contract in a separate place (as in the implementation
+    /// of [`SpreadLayout`]), so that it does not get overwritten by any contract
+    /// upgrade, which might introduce storage changes.
+    #[derive(Debug)]
+    #[cfg_attr(feature = "std", derive(ink_storage::traits::StorageLayout))]
+    struct ProxyFields {
+        /// The `Hash` of a contract code where any call that does not match a
+        /// selector of this contract is forwarded to.
+        delegate_to: Hash,
```
[the remark this answers is not in the document]
Comment (HCastano) [reply examples/upgradeable-contract/lib.rs:39]: Sorry I should've been more explicit here, I like the `forward_to` term more

**Review by HCastano — COMMENTED** [review]
So I think the biggest thing left is fixing any merge conflicts from #1165, and we should be good

File: crates/env/src/call/call_builder.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (xgreenx) [reply crates/env/src/call/call_builder.rs:82]: Then we can return `E::Balance` from `transferred_value`=) It is `u128`

[the remark this answers is not in the document]
Comment (HCastano) [reply crates/env/src/call/call_builder.rs:82]: Sure, but let's leave that for a follow up

**Review by HCastano — APPROVED** [review]
Okay, I think this is good to go.

There are some things we should improve in follow-ups, but this will at least let the community play with these new mechanisms. 

Thanks again Green!
